text/0000-additional-exported-layers.md

```diff
@@ -0,0 +1,123 @@
+
+# What it is
+[what-it-is]: #what-it-is
+
+Users would be able to define stack images with `VOLUME` directives in Dockerfiles.
+
+
+For example a build/run base image may look like - 
+
+```Dockerfile
+# Set a common base
+FROM ubuntu:bionic as base
+
+# Set required CNB information
+ENV CNB_USER_ID=1000
+ENV CNB_GROUP_ID=1000
+ENV CNB_STACK_ID="io.buildpacks.samples.stacks.bionic"
+LABEL io.buildpacks.stack.id="io.buildpacks.samples.stacks.bionic"
+
+# Create the user
+RUN groupadd cnb --gid ${CNB_GROUP_ID} && \
+  useradd --uid ${CNB_USER_ID} --gid ${CNB_GROUP_ID} -m -s /bin/bash cnb
+
+# Start a new run stage
+FROM base as run
+
+# Volume instructions to identify paths that the lifecycle
+# may safely overlay on the run image.
+# This would ensure that these paths are rebase safe as
+# docker currently discards any changes made to directories
+# declared as volumes. See https://docs.docker.com/engine/reference/builder/#notes-about-specifying-volumes
+VOLUME ["/app", "/opt/extensions", "/home/cnb/.config"]
+
+# Set user and group (as declared in base image)
+USER ${CNB_USER_ID}:${CNB_GROUP_ID}
+
+# Start a new build stage
+FROM base as build
+
+# Volume instruction in the build image that the lifecycle
+# will export out. This must match the volumes declared in the run image.
+VOLUME ["/app", "/opt/extensions", "/home/cnb/.config"]
+
+# Optionally users can override the default application workspace
+# by specifying it through the WORKDIR directive in the build
+# and run images.
+WORKDIR /app
+
+# Set user and group (as declared in base image)
+USER ${CNB_USER_ID}:${CNB_GROUP_ID}
+```
+
+Buildpacks will be passed a list of volumes through the `CNB_EXPORT_VOLUMES` variable and this variable would be a JSON list. Buildpacks may read this environment variable during `detect` or `build`.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+This RFC would require changes to the lifecycle and platform API.
+
+The platform would be responsible for mounting appropriate volumes based on the `Volumes` key in the OCI image config. For a platform like `pack` which relies on a daemon this should be fairly straight-forward to achieve since `docker run` automatically mounts appropriate volumes. For other platforms like `kpack` this would involve inspecting the builder image beforehand and modifying the build pod spec to accommodate the specified volumes.
+
+The lifecycle changes would involve exporting the files present in the above locations which should be similar to the logic that currently exists for exporting application workspace. Buildpacks could also take advantage of `slices` to specify paths in these additional directories that should exist as separate layers.
+
+Changes to the Buildpack API would be minimal and would mostly involve the Buildpacks being passed the additional `CNB_EXPORT_VOLUMES` variable for detection and build logic.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
```

Comment:
It would be hard for a simple platform like tekton to implement this.

Reply:
Is it possible this could open a new attack vector? I can't think of anything specific, but i wonder what's possible if a buildpack can drop files in `/etc/` or `/usr/bin/` for example.